app/controllers/api/v1/project_preferences_controller.rb

```diff
@@ -2,10 +2,29 @@ class Api::V1::ProjectPreferencesController < Api::ApiController
   include PreferencesController
 
   require_authentication :all, scopes: [:project]
+  resource_actions :create, :update, :show, :index, :update_settings
+  extra_schema_actions :update_settings
   schema_type :json_schema
+  before_action :validate_records, only: [:update_settings]
+
+  def update_settings
+    upp = UserProjectPreference.where(project_id: params_for[:project_id],
+                                      user_id: params_for[:user_id]).first
+    upp.settings.merge! params_for[:settings]
+    upp.save!
+    render status: :ok, nothing: true
+  end
 
   private
 
+  def validate_records
+    User.find params_for[:user_id]
```

Comment:
It's be good to not look these up unless we need to, instead of the current implementation we could use the `.find_by!` for the upp.

``` def upp_settings_update
  @upp = UserProjectPreference.find_by!(user_id: params_for[:user_id], project_id: params_for[:project_id])
  unless upp.project.owner?(api_user.user)
    raise ....
  end
```

And then access the upp via the instance param `@upp` in the update_settings action.